pkg/storegateway/bucket.go

```diff
@@ -615,102 +620,269 @@ func (s *BucketStore) Series(req *storepb.SeriesRequest, srv storepb.Store_Serie
+		seriesBuffer[i] = &storepb.StreamingSeries{}
+	}
+	seriesBatch := &storepb.StreamSeriesBatch{
+		Series: seriesBuffer[:0],
+	}
+	// TODO: can we send this in parallel while we start fetching the chunks below?
+	for seriesSet.Next() {
+		var lset labels.Labels
+		// IMPORTANT: do not retain the memory returned by seriesSet.At() beyond this loop cycle
+		// because the subsequent call to seriesSet.Next() may release it.
+		// TODO: check if it is safe to hold the lset.
+		lset, _ = seriesSet.At()
+
+		// We are re-using the slice for every batch this way.
+		seriesBatch.Series = seriesBatch.Series[:len(seriesBatch.Series)+1]
+		seriesBatch.Series[len(seriesBatch.Series)-1].Labels = mimirpb.FromLabelsToLabelAdapters(lset)
 
-			// Encode the message. We encode it ourselves into a PreparedMsg in order to measure
-			// the time it takes.
-			encodeBegin := time.Now()
+		// TODO: Add relevant trace spans and timers.
+
+		if len(seriesBatch.Series) == int(req.StreamingChunksBatchSize) {
 			msg := &grpc.PreparedMsg{}
-			if err = msg.Encode(srv, storepb.NewSeriesResponse(&series)); err != nil {
-				err = status.Error(codes.Internal, errors.Wrap(err, "encode series response").Error())
-				return
+			if err := msg.Encode(srv, storepb.NewStreamSeriesResponse(seriesBatch)); err != nil {
+				return status.Error(codes.Internal, errors.Wrap(err, "encode streaming series response").Error())
 			}
-			encodeDuration += time.Since(encodeBegin)
 
 			// Send the message.
-			sendBegin := time.Now()
-			if err = srv.SendMsg(msg); err != nil {
-				err = status.Error(codes.Unknown, errors.Wrap(err, "send series response").Error())
-				return
+			if err := srv.SendMsg(msg); err != nil {
+				return status.Error(codes.Unknown, errors.Wrap(err, "send streaming series response").Error())
 			}
-			sendDuration += time.Since(sendBegin)
-		}
-		if seriesSet.Err() != nil {
-			err = errors.Wrap(seriesSet.Err(), "expand series set")
-			return
+
+			seriesBatch.Series = seriesBatch.Series[:0]
 		}
+	}
+	if seriesSet.Err() != nil {
+		return errors.Wrap(seriesSet.Err(), "expand series set")
+	}
+
+	// We need to send hints and stats before sending the chunks.
+	// Also, these need to be sent before we send IsEndOfSeriesStream=true.
+	if err := s.sendHintsAndStats(srv, resHints, stats); err != nil {
+		return err
+	}
+
+	// Send any remaining series and signal that there are no more series.
+	msg := &grpc.PreparedMsg{}
+	seriesBatch.IsEndOfSeriesStream = true
+	if err := msg.Encode(srv, storepb.NewStreamSeriesResponse(seriesBatch)); err != nil {
+		return status.Error(codes.Internal, errors.Wrap(err, "encode streaming series response").Error())
+	}
+	// Send the message.
+	if err := srv.SendMsg(msg); err != nil {
+		return status.Error(codes.Unknown, errors.Wrap(err, "send streaming series response").Error())
+	}
+
+	if seriesSet.Err() != nil {
+		return errors.Wrap(seriesSet.Err(), "expand series set")
+	}
+
+	return nil
+}
+
+func (s *BucketStore) sendSeriesChunks(
```

Comment:
nitpick, but this can be just `sendChunks`

Reply:
In case we are not streaming, this sends both series and chunks. Do you still prefer sendChunks?